client/rpc/manual-xcm/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+use cumulus_primitives_core::ParaId;
+use futures::{future::BoxFuture, FutureExt as _};
+use jsonrpc_core::Result as RpcResult;
+use jsonrpc_derive::rpc;
+
+use parity_scale_codec::Encode;
+use xcm::latest::prelude::*;
+
+/// This RPC interface is used to manually submit XCM messages that will be injected into a
+/// parachain-enabled runtime. This allows testing XCM logic in a controlled way in integration
+/// tests.
+#[rpc(server)]
+pub trait ManualXcmApi {
+	/// Inject a downward xcm message - A message that comes from the relay chain.
+	/// You may provide an arbitrary message, or if you provide an emtpy byte array,
+	/// Then a default message (DOT transfer down to ALITH) will be injected
+	#[rpc(name = "xcm_injectDownwardMessage")]
+	fn inject_downward_message(&self, message: Vec<u8>) -> BoxFuture<'static, RpcResult<()>>;
+
+	/// Inject an HRMP message - A message that comes from a dedicated channel to a sibling
+	/// parachain.
+	///
+	/// Cumulus Parachain System seems to have a constraint that at most one hrmp message will be
+	/// sent on a channel per block. At least that's what this comment implies:
+	/// https://github.com/paritytech/cumulus/blob/c308c01b/pallets/parachain-system/src/lib.rs#L204
+	/// Neither this RPC, nor the mock inherent data provider make any attempt to enforce this
+	/// constraint. In fact, violating it may be useful for testing.
+	#[rpc(name = "xcm_injectHrmpMessage")]
+	fn inject_hrmp_message(
+		&self,
+		sender: ParaId,
+		message: Vec<u8>,
+	) -> BoxFuture<'static, RpcResult<()>>;
+}
+
+pub struct ManualXcm {
+	pub downward_message_channel: flume::Sender<Vec<u8>>,
+	pub hrmp_message_channel: flume::Sender<(ParaId, Vec<u8>)>,
+}
+
+impl ManualXcmApi for ManualXcm {
+	fn inject_downward_message(&self, msg: Vec<u8>) -> BoxFuture<'static, RpcResult<()>> {
+		let downward_message_channel = self.downward_message_channel.clone();
+		async move {
+			// If no message is supplied, inject a default one.
+			let msg = if msg.is_empty() {
```

Comment:
What if we make msg an Option here? I feel it looks better fro the rpc method too, what do you think?

Reply:
To summarize our discussion in slack:

* Using the empty message as a way to trigger the default message is definitely a sloppy/hacky thing to do in a language with things like `Option`.
* One nice property that we'd like to keep is that you can send a valid encoded xcm directly to the RPC with no further encoding. `Option` doesn't do that.
* We could have a separate RPC for this message
* Or a separate endpoint for downward transfers and others
* We don't want to over-design it before we know how it will be used.

All that considered, we will stick with the current approach.